OpenSim/Common/Component.cpp

```diff
@@ -838,26 +836,61 @@ void Component::constructOutputForStateVariable(const std::string& name)
             SimTK::Stage::Model);
 }
 
+// mark components owned as properties as subcomponents
+void Component::markPropertiesAsSubcomponents()
+{
+    // if being invoked either constructing a new Component
+    // or the properties have been modified. In the latter case
+    // we must make sure that pointers to old properties are cleared
+    _components.clear();
+
+    // Now mark properties that are Components as subcomponents
+    for (int i = 0; i < getNumProperties(); ++i) {
+        auto& prop = updPropertyByIndex(i);
+        if (prop.isObjectProperty()) {
+            for (int j = 0; j < prop.size(); ++j) {
+                Object& obj = prop.updValueAsObject(j);
+                if (Component* comp = dynamic_cast<Component*>(&obj) ) {
+                    markAsSubcomponent(comp);
+                }
+                //Set<Component>* objects = dynamic_cast<Set<Component>*>(&obj)
+                else {
+                    std::string objType = obj.getConcreteClassName();
+                    if (obj.hasProperty("objects")) {
+                        auto& objectsProp = obj.updPropertyByName("objects");
+                        for (int k = 0; k < objectsProp.size(); ++k) {
+                            Object& obj = objectsProp.updValueAsObject(k);
+                            if (Component* comp = dynamic_cast<Component*>(&obj) )
+                                markAsSubcomponent(comp);
```

Comment:
Hm, not sure I understand completely. Here's my question more precisely:

Line 853 checks to see whether the property is a Component; if it is, then mark it as a subcomponent. Otherwise, line 859 checks to see whether the property might have Components buried within it, and line 863 (which is identical to line 853) checks for more Components at that level. But what if there are Components buried under _two_ layers of the `hasProperty("objects")` test? Is it not necessary to call lines 857 to 864 recursively in the event that there are Components many layers deep?